Handle more arithmetic operators for native histograms

zenador opened the pull request:
Addresses https://github.com/prometheus/prometheus/issues/12250

Works in unit tests and when manually testing on the UI with scalars and float metrics. This fixes the issue we observed with query sharding histograms with instant query splitting, however the float imprecision accumulates and the final result is very different in unit tests (tolerance was originally 1e-12 for floats but the histogram ones only passed with 1e-2).

Also added avg for native histograms as in https://github.com/prometheus/prometheus/issues/11973, to be discussed there.

krajorama commented on promql/engine.go:
```diff
@@ -2419,10 +2421,27 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
+		}
 		return lhs * rhs, nil, true
 	case parser.DIV:
+		if hlhs != nil && hrhs == nil && rhs != 0 {
```
I think we can probably allow rhs==0 and let it recover like the float series case

beorn7 commented on promql/engine.go:
```diff
@@ -2583,46 +2607,77 @@ func (ev *evaluator) aggregation(op parser.ItemType, grouping []string, without
 			if s.H != nil {
 				group.hasHistogram = true
+				if group.hasFloat && group.hasHistogram {
+					break
```
I'm not sure if that additional check here is really helping. The case of mixed aggregations should be very rare. By adding this check here, we are creating overhead for the common case and get just a bit of performance back for the rare error case.

beorn7 commented on promql/engine.go:
```diff
@@ -2583,46 +2607,77 @@ func (ev *evaluator) aggregation(op parser.ItemType, grouping []string, without
 			} else {
 				group.hasFloat = true
+				if group.hasFloat && group.hasHistogram {
+					break
```
Again, I don't think this condition is beneficial in the big picture.

beorn7 reviewed (changes requested):
Looks great. Just some not-so-substantial comments.

Special thanks for the testing. This should be much easier once #11170 is done.

Two other things to do:
1. Amend the documentation to explain what works now.
2. I think we can/should also fix other aggregation operators (stddev etc.). Some of them can be made work with histograms, others should ignore them cleanly (currently, we count them, which might throw some results off).

I think (1) needs to be included in this PR, but (2) can be done in a separate PR.

zenador commented:
Besides what you mentioned, noticed a few more issues:
- `Mul` is an exact duplicate of `Scale` which is used for `rate`. Removed `Mul`. Please let me know if you prefer to rename all the `Scale`s to `Mul` or something.
- When adding, subtracting, and aggregating, we do not compact the histogram, ending up with ugly redundant buckets (but this problem doesn't happen with sum/avg_over_time, so we had a lot of redundancy in `TestNativeHistogram_Sum_Count_Add_AvgOperator`, with some expected histograms compacted and some not. I added the compact step to fix that, but not sure if this is unnecessary overhead if we don't actually care if the result is compacted.

beorn7 commented:
Good catch with the `Scale` vs `Mul`. Now that we have `Div` and `Add` and `Sub`, maybe we should call `Scale` `Mul` indeed.

I'll get to the other points ASAP.

beorn7 reviewed (approved):
Thank you very much. That's a huge chunk of work done.

I hope it's OK if I squash the 14 commits into one.
